# Remove during next Triton integrate. This fix was cherry-picked from upstream:
# https://github.com/triton-lang/triton/commit/35302c4747e1d1c4f9a8a94b1870ebca2d26ed18

--- a/lib/Conversion/TritonGPUToLLVM/WarpSpecializeUtility.cpp
+++ b/lib/Conversion/TritonGPUToLLVM/WarpSpecializeUtility.cpp
@@ -72,8 +72,11 @@
       return WalkResult::skip();
     }
     if (barrierHelper.isBarrierOp(op)) {
-      return WalkResult(lowerBarrier(op, defaultNumWarps, /*partitionIdx=*/{},
-                                     barrierHelper));
+      auto barRes =
+          lowerBarrier(op, defaultNumWarps, /*partitionIdx=*/{}, barrierHelper);
+      // Since this is a PreOrder walk, we have to return skip() on success so
+      // we do not attempt to visit the regions in the now deleted op.
+      return barRes.succeeded() ? WalkResult::skip() : WalkResult::interrupt();
     }
     if (auto callOp = dyn_cast<LLVM::CallOp>(op)) {
       if (!innerFunctions.contains(callOp.getCalleeAttr().getAttr()))
